GP11 : TEE_SetOperationKey/2 panic reason

cedric-chaumont-st-dev opened the pull request:
Signed-off-by: Cedric Chaumont <cedric.chaumont@st.com>

jbech-linaro commented on lib/libutee/tee_api_operations.c:
```diff
@@ -444,18 +451,24 @@ TEE_Result TEE_SetOperationKey2(TEE_OperationHandle operation,
+		goto out;
+	}
 
 	/* Two keys not expected */
```
I can see that this hasn't been changed by you, but is this really correct? Reading the spec it says:
```
TEE_HANDLE_FLAG_EXPECT_TWO_KEYS : Set if the algorithm expects two keys to be set, using
TEE_SetOperationKey2 . This happens only if algorithm is set to TEE_ALG_AES_XTS .
```

And in `TEE_SetOperationKey2` one can read:
```
The TEE_SetOperationKey2 function initializes an existing operation with two keys. This is used only for
the algorithm TEE_ALG_AES_XTS .
```

jbech-linaro commented:
Just one comment, other than that:
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

cedric-chaumont-st-dev commented:
Updated.

cedric-chaumont-st-dev commented:
Updated.

pascal-brand-st-dev commented on lib/libutee/tee_api_operations.c:
```diff
@@ -464,97 +476,119 @@ TEE_Result TEE_SetOperationKey2(TEE_OperationHandle operation,
-		     requiredKeyUsage) != operation->info.requiredKeyUsage) {
-			TEE_Panic(0);
-		}
+	res = TEE_GetObjectInfo1(key1, &key_info1);
```
Here, `key1` can be `TEE_HANDLE_NULL`. So it may be wrong I think.

cedric-chaumont-st-dev replied:
@pbrand: the key1/key2 check should prevent from that (panic). what do you think ?

seungeun2 replied:
It looks like when key1 key2 are both TEE_HANDLE_NULL, it leads to panic(0) at the end of the function. But spec specifies one of panic reasons as follows.
 "key1 and key2 are not both TEE_HANDLE_NULL and key1 or key2 or both are not valid handles on a key object."
To me, it looks like the function shouldn't panic when both key1 and key2 are TEE_HANDLE_NULL.

cedric-chaumont-st-dev commented:
Rebased. Updated.

pascal-brand-st-dev commented:
Reviewed-by: Pascal Brand <pascal.brand@linaro.org>

cedric-chaumont-st-dev commented:
Squashed. Rebased. Tested (STM, ARM Juno).
28701 subtests of which 0 failed
687 test cases of which 0 failed
0 test case was skipped
TEE test application done!

seungeun2 commented:
It looks like when key1 key2 are both TEE_HANDLE_NULL, it leads to panic(0) at the end of the function. But spec specifies one of panic reasons as follows.
"key1 and key2 are not both TEE_HANDLE_NULL and key1 or key2 or both are not valid handles on a key object."
To me, it looks like the function shouldn't panic when both key1 and key2 are TEE_HANDLE_NULL.

cedric-chaumont-st-dev commented:
"• key1 and/or key2 are not initialized." : both  key1 and  key2 have to be initialized (different from TEE_HANDLE_NULL) otherwise Panic.
"• key1 and key2 are not both TEE_HANDLE_NULL and key1 or key2 or both are not valid handles on a key object" : key1 and key 2 are both initialized (different from TEE_HANDLE_NULL) but both key1 or key2 have to be a valid handle (from TEE_GetObjectInfo1) otherwise Panic. I hope it helps.

seungeun2 commented:
As far as I understand, "key1 and key2 are not both TEE_HANDLE_NULL" (which @cedric-chaumont-st-dev  paraphrased as "key1 and/or key2 are not initialized") corresponds to the case where key1 and key2 are not both TEE_HANDLE_NULL, i.e., either 'only key1 is TEE_HANDLE_NULL' or  'only key2 is TEE_HANDLE_NULL'.  The spec doesn't say anything about the case when both key1 and key2 are TEE_HANDLE_NULL.(I'm not a native English speaker, so.. correct me if i'm wrong) 

The same thing goes to TEE_SetOperationKey function. The spec specifies one of panic reasons as "key is not TEE_HANDLE_NULL and is not a valid handle on a key object". It doesn't say that TA needs to panic when key is equal to TEE_HANDLE_NULL.  But it says "If key is set to TEE_HANDLE_NULL , then the operation key is cleared."  which is applied well in this PR.
But what's clearing the operation key for once TA panics?

cedric-chaumont-st-dev commented:
@seungeun2: 
"• key1 and/or key2 are not initialized." is part of the Gp internal core API specification v1.1 . if a key is not initialized then it is equal to TEE_HANDLE_NULL. If one of the 2 keys or both are equal to TEE_HANDLE_NULL then it must Panic. In any case, if one of the 2 keys is equal TEE_HANDLE_NULL it will Panic and it ensures both of them are initialized at the same time.

"But what's clearing the operation key for once TA panics?" . Could you rephrase your question please ?

cedric-chaumont-st-dev commented:
There is something confusing in the spec:
"key1 and key2 MUST both be non-NULL or both NULL" vs "key1 and/or key2 are not initialized".
I am going to clarify with GP WG.

seungeun2 commented:
@cedric-chaumont-st-dev : 
What I meant by "But what's clearing the operation key for once TA panics?" is that when key is TEE_HANDLE_NULL, the operation key needs to be cleared right? Buf if TA panics, there's no point in clearing the operation key , cause TA's not available to use the operation~ This is why  I asked the question "But what's clearing the operation key for once TA panics?" and I think TA shouldn't panic when key is TEE_HANDLE_NULL.
I would really appreciate if you keep me updated after you clarify with GP WG.
Thank you
